### DISCUSSION TITLE
如何在 dwl 里用 gdb 调试 fcitx5?

### DISCUSSION THREAD

**Discussion by guyuming76** [fcitx/fcitx5#551] [Q&A] [answered]
我读了 https://www.csslayer.info/wordpress/fcitx-dev/fcitx-5-debug/

之前照着在 KDE plasma konsole 里面用gdb 调试fcitx5 是可以的，记了笔记 https://gitee.com/guyuming76/personal/commit/209d33bf308188138f05685853def1bdd232bf10

但今天，我尝试在 DWL (wayland) 里从 alacritty, foot 里启动 gdb 调试fcitx5, 那些个如gtk_im_module 变量设置似乎没有用，进入gdb后，键盘还是卡死了，我可以通过自己的日志看到按键发送给了 IM ，但IM没有回应。我应该如何设置？

如果我需要修改啥代码，能够有些思路指导吗？

![20220723_08h38m48s_grim](https://user-images.githubusercontent.com/12045548/180585992-d10fa038-e7bd-45cc-9743-202368f7f808.png)

**Comment by wengxt** [comment, marked as answer]
当然没有用，因为alacritty 和 foot 也不是用 gtk 啊，怎么会有用呢？
具体你说的这个问题，我只能建议不要用 alacritty 和 foot，因为他们的输入法支持是基于 wayland 的，而 wayland 的输入法会在激活的时候输入法会创建一个 keyboard grab，这是 compositor 层面的，而不是程序层面的，确实没办法。

建议用 xterm。

**Reply by wengxt** [reply]
另一个建议是，给你的 dwl 实现一个 nested 模式，就像你可以在 weston 里再启动一个 weston 一样（X11的例子就像是 Xephyr），然后调试 nested 这个，而不是你正在用于显示的这个

**Reply by guyuming76** [reply]
谢谢，我试下。我试过在另一个 tty （ctrl+alt+F2）里startx 启动一个DWM,然后从里面用gdb,可以调试 fctix5, 但特别卡，就是说dwl里面拼音输入法卡住了，以为是击中断点了，但换到 dwm 里的 gdb ，看到还是 continue 状态，有时隔了好久，看到gdb里已经停在断点了。

更新：关于卡这件事，我刚才发现，可能和gdb调试无关，我从 DWL 里用 ctrl+alt+f2 切换到 DWM 里，会发现 DWM 显示的是前一个屏幕，过了一会儿又自动跳到正确的屏幕上了，这个不太正常，我之前没注意到有这种现象。

更新：题外话，关于卡这件事，我觉得可能是alacritty的原因，切换至的DWM的tag里运行了alacritty和ranger. 刚切过来的一段时间，键盘老是不听使唤。但我在同一个tag上运行个st就没这个问题，且这段时间dwm的键比如MOD+enter啥的是好用的。 

另外，既然到另一个tty运行gdb,其实我也不一定要用dwm,直接tty里gdb调试也行，只是我还没尝试过给tty framebuffer截屏。

更新 2022/7/31： 今天在自己的gnu/linux 学习笔记里面看到下面这行，我竟不记得自己用过！
安装 media-gfx/fbgrab 后，可以 sudo fbgrab 文件名.png 在tty 里面通过framebuffer截屏

**Comment by guyuming76** [comment]
我试了下用DWM,里面套了个DWL，然后用dwm 里面的gdb attach dwl 里的 fctix5, 是可以调试的。

DWM套DWM 或 DWL套DWL 可能不行，因为ModKey设一个外面的传不进去，dwm和dwl的ModKey设不同就行。

![debugFcitx5](https://user-images.githubusercontent.com/12045548/180598132-e98d7f86-9716-4a78-93dd-51b1b10ce5de.png)

顺便说下，我想调一下fcitx5是想研究下那个setScaleAndTransform 相关内容。 我在投影仪上看屏幕，有时中文输入太小，看不清，alacritty等VT可以用ctrl + 放大，但是输入法选词框貌似大小不变。如果有可能增强我是不是要另外提一个功能增强需求？

更新：不过这个快捷键的确不那么容易设置，alacritty里ctrl +放大，st 里面貌似是ctrl shift +, 就是不同应用放大字体的快捷键都不一样。单独为输入法设计一个又不容易记。貌似 fcitx5-configtools 里面 classic ui addon 设置里面的字体设置有用的，但是不是要重启fcitx后才起作用呢？

**Reply by wengxt** [reply]
你应该让你的dwl支持设置屏幕的scale

**Reply by guyuming76** [reply]
多谢！ dwl 和scale 设置不冲突，我用了下面命令可以放大屏幕字体
wlr-randr --output VGA-1 --scale 1.5
嫌命令麻烦就在waybar里做个支持点击的标签就行了，如我上面图中的screenshot, sleep.

如果是在DWM(xorg)里用
xrandr --output VAG1 --scale 0.5x0.5
